### DISCUSSION TITLE
Non-auto disposed provider vs auto disposed with maintainState = true

### DISCUSSION THREAD

**Discussion by wujek-srujek** [rrousselGit/riverpod#660] [Q&A] [answered]
What is the difference between a non-'auto disposed' provider vs an 'auto disposed' one with `maintainState` set to `true`? It seems to me that they do pretty much the same thing. When debugging, the callback passed to `ref.onDispose` is never called when `maintainState = true`, i.e. the same as simply not making the provider 'auto disposed'. What am I missing?

**Comment by rrousselGit** [comment, marked as answer]
In terms of behaviour, they are similar.

The real difference is in terms of debugging.

It is purposefully not allowed to read an auto-dispose provider from a non-auto-disposed to avoid mistakes.
Whereas an auto-dispose provider with `maintainState = true` does not prevent those mistakes.

**Reply by wujek-srujek** [reply]
So an auto-dispose provider with `maintainState = true` will behave like a non-auto-dispose provider except that additionally it allows me to read other auto-disposed provider, possibly creating a leak, right?

This is pretty confusing, this should be documented somewhere - maybe it is already and I missed it?

**Reply by rrousselGit** [reply]
There is no leak involved.

I don't see what's confusing.

What were expecting?

**Reply by wujek-srujek** [reply]
Of course you don't see anything confusing, this is your code and you know this like the back of your hand ;) But most of us mere mortals don't have this privilege. The fact that I started this discussion means it is not obvious to at least one person, and I did read the docs and some of the comments in code.

What is confusing is that there are two concepts (auto-dispose with maintainState = true and no-auto-dispose) which result in similar/the same behavior, without any documentation or explanation what the actual difference is.

1. How should users know which one to choose and when?
2. How should users know what the difference is?
3. If the difference is in terms of debugging, as you said, what do you actually mean? How does it aid me in debugging?

By 'leak' I don't mean a memory leak in the most common form; I mean an auto-dispose provider being kept alive longer than otherwise necessary, which would happen if you allowed a non-auto-dispose provider to read/watch an auto-disposed one, right? (Or when an auto-disposed provider with maintainState = true uses an auto-disposed provider, as I learned here.)

**Reply by mono0926** [reply]
@rrousselGit 

> It is purposefully not allowed to read an auto-dispose provider from a non-auto-disposed to avoid mistakes.

Do you mean not allowed to `watch`, not `read`?

**Reply by wujek-srujek** [reply]
Yeah, you can't `watch` or `listen` in this case, `read` is fine as it doesn't register any listener.

**Reply by iamarnas** [reply]
My explanation. 

> How should users know which one to choose and when?

Always start use non auto dispose providers. Auto dispose providers are build most for a asynchronous data handling example for `FutureProvider` or `StreamProvider`. And if you are forced combine them with others providers then you will be forced to convert them to auto dispose provider also. In this case you simple `Provider` can be `Provider.autoDispose( )`.

> How should users know what the difference is?

Auto dispose providers have `maintainState`. The `maintainState` allows the provider to tell Riverpod if the state of the provider should be preserved even if no-longer listened. [Example](https://riverpod.dev/docs/concepts/modifiers/auto_dispose/#refmaintainstate):

```dart
final myProvider = FutureProvider.autoDispose((ref) async {
  final response = await dio.get(...);
  ref.maintainState = true;
  return response;
});
```
This way, if the request failed and the user leaves the screen then re-enters it, then the request will be performed again. But if the request completed successfuly, **_the state will be preserved and re-entering the screen will not trigger a new request._**

So user should use simple providers as standard. Most of them i see just wrap everything to auto dispose even YouTube  experter do that 😁 and they think it's safer because `auto` word sounds safer . Flutter dispose them both in all case and only `maintainState` make difference as in the example. If you don't need save request calls or just loading from the local data, then you don't need auto dispose at all.

**Reply by wujek-srujek** [reply]
I have to admit I'm not convinced with your explanations.

I'm one of those people who (possibly wrongly?) slap 'autoDispsose' on everything. For some reason I had the impression that this is recommended, I got that impression by reading the official docs (again, the impression is possibly wrong).

As for 'autoDisposed' being mainly for async - the docs say (https://riverpod.dev/docs/concepts/modifiers/auto_dispose):
```
There are multiple reasons for doing such, such as:
...
- To reset the state when the user leaves a screen and re-enters it.
```
This is a very common use case that has nothing to do with asynchrony in my opinion. For example, I have a page which allows the user to input complex data, and all of it is managed by riverpod - when the user goes back, I want to simply forget everything, and when the user re-enters the screen, the whole things starts over with no previous state. Without 'autoDispose' I would need to reset manually, which is simply not as comfortable.

I also don't think Flutter will dispose a non-autoDipsosed provider as you say - once added to the riverpod container, it will stay there, right? That's because the container is not going out of scope as it is one of the topmost widgets most of the time. And once in the container, such provider will be available to whoever has access to the provider variable, and in my case re-entering the page would show the previous state, as mentioned above.

**Reply by iamarnas** [reply]
> I also don't think Flutter will dispose a non-autoDipsosed provider.

Have you tried to debug it and compare?

There is nothing wrong with using only the auto dispose providers, it is up to you. Even if you don't know when to use auto dispose you never will be wrong. `maintainState` you can use for all types of providers, it only preserves the state and not just for asynchronous providers. I do not know why Remi does not do it by default and removes `.autoDispose()` so as not to complicate beginners.

**Reply by wujek-srujek** [reply]
Yes, I have debugged it that's why I know how it works. But I might have made mistakes and I did it with an older version, I might have to do it again.

```
maintainState you can use for all types of providers
```
No, `maintainState` cannot be used for all types of providers, it isn't available for non-auto-disposed at all, the code won't even compile: https://github.com/rrousselGit/river_pod/blob/6ce875ae62b325f0f5af6d96da2bad9196fd826f/packages/riverpod/lib/src/framework/auto_dispose.dart#L6

Anyways, the whole point of the discussion is that I (and others I have spoken to) don't know when to use auto-disposed with `maintainState = true` vs simply non-auto-disposed, which are pretty much the same according to Remi. And this discussion hasn't been very helpful so far, unfortunately.

**Reply by rrousselGit** [reply]
The answer is straightforward: If a non-auto-dispose provider fits you, use that.

**Reply by iamarnas** [reply]
> `maintainState` you can use for all types of providers.

I meant `Provider.autoDispose()`, `StateProvider.autoDispose()`, `StateNotifierProvider.autoDispose()` and...

**Reply by rrousselGit** [reply]
In general, use `maintainState` only if you want to _dynamically_ prevent dispose. Like the doc mentions with https://riverpod.dev/docs/concepts/modifiers/auto_dispose#refmaintainstate

This example maintains the state only if a network requests succeeds.

**Reply by wujek-srujek** [reply]
> > I also don't think Flutter will dispose a non-autoDipsosed provider.
> 
> Have you tried to debug it and compare?

@iamarnas just for completeness, I have just checked with the latest riverpod version - PageA sets a state in a non-auto-disposed provider and pushes PageB; when PageB is popped, the state in the provider is still there and not disposed. This is exactly the use case for auto-disposing and I don't understand why you said Flutter would dispose it anyway; if it did, what would be the added value of auto-disposed?

**Reply by iamarnas** [reply]
@wujek-srujek It not dispose because `ProviderScope` is on the top widget three and keep the all providers state.

**Reply by wujek-srujek** [reply]
Yes, this is what I meant by this:

> I also don't think Flutter will dispose a non-autoDipsosed provider as you say - once added to the riverpod container, it will stay there, right? That's because the container is not going out of scope as it is one of the topmost widgets most of the time. And once in the container, such provider will be available to whoever has access to the provider variable, and in my case re-entering the page would show the previous state, as mentioned above.

to which you said that I should test it, so I did. Many misunderstandings, we have.
